README.md
Corsproxy<img width="18%" align="right" src=".graphics/corsproxy-logo.svg">
=========

This is a a simple CORS proxy server suitable to install as a system service on Linux servers.  It uses [CORS Anywhere](https://github.com/Rob--W/cors-anywhere).

[![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg?style=flat-square)](https://choosealicense.com/licenses/bsd-3-clause)
[![Latest release](https://img.shields.io/github/v/release/caltechlibrary/corsproxy.svg?style=flat-square&color=b44e88)](https://github.com/caltechlibrary/corsproxy/releases)


Table of contents
-----------------

* [Introduction](#introduction)
* [Installation](#installation)
* [Usage](#usage)
* [Known issues and limitations](#known-issues-and-limitations)
* [Getting help](#getting-help)
* [License](#license)
* [Authors and history](#authors-and-history)
* [Acknowledgments](#authors-and-acknowledgments)


Introduction
------------

The development of web-based applications, particularly single-page applications written using JavaScript, can be stymied by problems involving [CORS](https://en.wikipedia.org/wiki/Cross-origin_resource_sharing) security measures enforced by web browsers.  One problem happens when a network server providing a remote API service does not support CORS: if the nature of the network API requires nontrivial types of operations (e.g., HTTP POST requests that contain data payloads), the web browser running the single-page application will enforce CORS requirements, and the API requests will fail when the server does not respond correctly.

A simple solution to this problem is to insert an intermediate proxy server between the web application and the network service.  An example of such a proxy server is [CORS Anywhere](https://cors-anywhere.herokuapp.com/), an open-source proxy server that runs in NodeJS.  CORS&nbsp;Anywhere works well, and only needs some additional elements to make it suitable for running as a standard system service on a Linux server.  This project (Corsproxy) aims to provide those additional elements.  Corsproxy also tries to simplify some of the configuration steps in using CORS Anywhere.


Installation
------------

Detailed installation and configuration instructions are given in the file [admin/README](admin/README.md).


Usage
-----

Corsproxy is meant to be started and stopped using standard system tools such as `systemctl` on CentOS.  Detailed instructions are given in the file [admin/README](admin/README.md).


Known issues and limitations
----------------------------

* The way that host restrictions/rate limits are implemented is based primarily on the Origin header in HTTP requests.  This works and has some security advantages, but makes it more difficult to configure in some other situations.


Getting help
------------

If you find an issue, please submit it in [the GitHub issue tracker](https://github.com/caltechlibrary/corsproxy/issues) for this repository.


License
-------

Software produced by the Caltech Library is Copyright (C) 2020, Caltech.  This software is freely distributed under a BSD/MIT type license.  Please see the [LICENSE](LICENSE) file for more information.


Authors and history
---------------------------

Michael Hucka developed the first version of Corsproxy after running into problems during the implementation of a Vue.js based application that had to interact with a non-CORS enabled network service.

[Erik Demaine](https://github.com/edemaine) contributed a pull request in Feb. 20201 to add support for explicit white lists and black lists.


Acknowledgments
---------------

Corsproxy makes use of [CORS Anywhere](https://github.com/Rob--W/cors-anywhere), without which it would have been effectively impossible to develop Corsproxy with the resources we had.

The [vector artwork](https://thenounproject.com/term/subtotal/2820924/) used as a starting point for the logo for this repository was created by [Timofey Rostilov ](https://thenounproject.com/t.rostilov/) for the [Noun Project](https://thenounproject.com).  It is licensed under the Creative Commons [Attribution 3.0 Unported](https://creativecommons.org/licenses/by/3.0/deed.en) license.  The vector graphics was modified by Mike Hucka to change the color and remove text.

This work was funded by the California Institute of Technology Library.

<div align="center">
  <br>
  <a href="https://www.caltech.edu">
    <img width="100" height="100" src=".graphics/caltech-round.png">
  </a>
</div>


CHANGES.md
Change log for REPOSITORY
=========================

Version 1.1.0
-------------

* Merge PR #1 from [Erik Demaine](https://github.com/edemaine) to add support for black lists and white lists.
* Fix [security vulnerability in `http-proxy` module](https://github.com/advisories/GHSA-6x33-pw7p-hmpq) by updating dependency to version 1.18.1.
* Fix out-of-date links in supporting info files.


Version 1.0.1
-------------

The only change in this release is to rename the main JavaScript program file to corsproxy.js, instead of server.js, to make it more obvious and easier to find in a process list.


Version 1.0.0
-------------

This release adds support for HTTPS. The bump in version number also represents the fact that this software has been running without problems on a live server used for internal Caltech Library work, and thus it seems stable enough to call this a proper 1.0.0 release.


admin/server.sh
#! /bin/bash
## ============================================================================
## File name   : server.sh
## Description : server control script for cors-proxy
## Author(s)   : Michael Hucka <mhucka@caltech.edu>
## Organization: California Institute of Technology
## Date created: 2020-04-23
## ============================================================================

# Preliminary definitions
# .............................................................................

# This definition is used in a pipe command in the main section.
# Don't merge this with the next function definition.
LOG="logger -s -t corsproxy -p daemon.info"

log() {
    $LOG $1 >&2
}


# Read configuration file and set defaults.
# .............................................................................

if [[ -e config.sh ]]; then
    log "Reading configuration file"
    source config.sh
else
    # Default values in case config.sh is missing.
    PID_FILE=/tmp/corsproxy.pid
    RESTART_PAUSE=10
fi


# Main code.
# .............................................................................

case "$1" in
    start)
        log "Starting server process from directory `pwd`"
        # Problem: we need the PID from the first process in the pipe, but if
        # you use $! you'll get the PID of the last process in the pipe.
        # Solution: writes the PID to file descriptor 3, so that it can be
        # read from there on the following line.  Genius idea posted by user
        # John Kugelman, 2014-07-16, https://stackoverflow.com/a/3786955/743730
        TMPFILE=$(mktemp /tmp/corsproxy.XXXXXX)
        trap 'rm -f -- "$TMPFILE"' INT TERM HUP EXIT
        ( node ../server/server.js 2>&1 & echo $! >&3) 3>$TMPFILE | $LOG & >&2
        sleep 2
        PID=$(<$TMPFILE)
        if [[ -z $PID ]]; then
            log "Failed to start server process" >&2
            exit 1
        else
            echo $PID > $PID_FILE
            log "Server started successfully with PID $PID." >&2
        fi
        exit 0
        ;;

    stop)
        if [[ ! -e $PID_FILE ]]; then
            log "PID file $PID_FILE does not exist." >&2
            exit 2
        fi
        PID=$(<$PID_FILE)
        log "Sending kill signal to process $PID" >&2
        if ! kill -TERM  $PID> /dev/null 2>&1 ; then
            # We don't want to start killing processes by name, because the
            # user might have more than one server running.  Just give up.
            if ps -p $PID > /dev/null 2>&1 ; then
                log "Could not terminate process $PID." >&2
                exit 1
            else
                log "Process $PID no longer exists." >&2
                exit 1
            fi
        else
            log "Process $PID killed." >&2
            exit 0
        fi
        ;;

    restart)
        $0 stop
        if [[ $? -eq 0 ]]; then
            log "Pausing for $RESTART_PAUSE s." >&2
            sleep $RESTART_PAUSE
        fi
        log "Restarting server." >&2
        $0 start
        exit $?
        ;;

    status)
        if [[ ! -e $PID_FILE ]]; then
            log "PID file $PID_FILE does not exist." >&2
            exit 2
        fi
        PID=`cat $PID_FILE`
        if ps -p $PID > /dev/null 2>&1 ; then
            log "Process $PID running." >&2
            exit 0
        else
            log "Process $PID no longer exists." >&2
            exit 1
        fi
        ;;

    *)
        echo "Usage: $0 {start|stop|restart|status}." >&2
        exit 3
        ;;
esac


server/server.js
// ============================================================================
// File name   : server.js
// Description : CORS proxy server main entry point.
// Author(s)   : Michael Hucka <mhucka@caltech.edu>
// Organization: California Institute of Technology
// Date created: 2020-04-23
//
// This is a significantly modified version of server.js from cors-anywhere,
// originally obtained from https://github.com/Rob--W/cors-anywhere
// ============================================================================

// Requirements.
// ............................................................................

var fs             = require('fs');
var path           = require("path");
var cors_anywhere  = require('./node_modules/cors-anywhere/lib/cors-anywhere');
var checkRateLimit = require('./node_modules/cors-anywhere/lib/rate-limit');


// Environment variables used for run-time configuration.
// ............................................................................
// These are set using a config file associated with our server process
// start/stop/restart script.  More info about the meaning of these variables
// can be found in the CORS Anywhere docs.

// Listen on a specific host via the HOST environment variable.
var host = process.env.HOST || '0.0.0.0';

// Listen on a specific port via the PORT environment variable.
var port = process.env.PORT || 8080;

// HTTPS key file
var key_file = process.env.KEY_FILE || '';

// HTTPS certificate file
var cert_file = process.env.CERT_FILE || '';

// The rate limit sets both the default number of connections allowed from
// unlisted hosts, and a list of hosts that are allowed unlimited connections.
// The value of RATELIMIT must be a list of the form "N X host [host ...]",
// where the first 2 numbers indicate the rate limit for unknown origins, and
// the list of hosts is the whitelist (hosts without limits).  The numbers N
// and X indicate N requests per X minutes for connections that are *not*
// coming from the listed hosts, Hosts can be regular expressions of the
// JavaScript variety, set off by forward slashes.  For example, "0 1
// /.*\.caltech\.edu/" means 0 connections per minute are allowed by default,
// except for conections from *.caltech.edu.  Note that this only affects
// actual CORS proxying; connections to the help page are *not* affected by
// this at all, and are unrestricted.  (Corrolary: don't use connections to
// the help page to judge whether this setting is working as desired.)
var rateLimit = process.env.RATELIMIT || '30 1';

// Header that must be present in HTTP request in order for corsproxy to
// accept it.
var requiredHeader = process.env.REQUIRED_HEADER || '';

// File with text to override the default CORS Anywhere help text.
var helpTextFile = path.join(__dirname, 'help-text.txt');


// Main code.
// ............................................................................

process.on('SIGTERM', () => {
    console.log('Received SIGTERM; exiting.');
    process.exit(0);
});

if (key_file && cert_file) {
    console.log('Using HTTPS certificate', cert_file);
    httpsOptions = {
        key: fs.readFileSync(key_file),
        cert: fs.readFileSync(cert_file)
    };
} else
    httpsOptions = {};


try {
    cors_anywhere.createServer({
        requireHeader      : requiredHeader,
        setHeaders         : {"X-Proxied-by": "CORS Proxy"},
        checkRateLimit     : checkRateLimit(rateLimit),
        redirectSameOrigin : true,
        httpsOptions       : httpsOptions,
        helpFile           : helpTextFile,
    }).listen(port, host, function() {
        console.log('Running CORS Proxy on ' + host + ':' + port);
        console.log('Ratelimit configuration: "' + rateLimit + '"');
        console.log('Required header: ' + requiredHeader);
    });
} catch (err) {
    console.log(err);
    process.exit(1);
}


server/corsproxy.js
// ============================================================================
// File name   : server.js
// Description : CORS proxy server main entry point.
// Author(s)   : Michael Hucka <mhucka@caltech.edu>
// Organization: California Institute of Technology
// Date created: 2020-04-23
//
// This is a significantly modified version of server.js from cors-anywhere,
// originally obtained from https://github.com/Rob--W/cors-anywhere
// ============================================================================

// Requirements.
// ............................................................................

var fs             = require('fs');
var path           = require("path");
var cors_anywhere  = require('./node_modules/cors-anywhere/lib/cors-anywhere');
var checkRateLimit = require('./node_modules/cors-anywhere/lib/rate-limit');


// Environment variables used for run-time configuration.
// ............................................................................
// These are set using a config file associated with our server process
// start/stop/restart script.  More info about the meaning of these variables
// can be found in the CORS Anywhere docs.

// Listen on a specific host via the HOST environment variable.
var host = process.env.HOST || '0.0.0.0';

// Listen on a specific port via the PORT environment variable.
var port = process.env.PORT || 8080;

// HTTPS key file
var key_file = process.env.KEY_FILE || '';

// HTTPS certificate file
var cert_file = process.env.CERT_FILE || '';

// The rate limit sets both the default number of connections allowed from
// unlisted hosts, and a list of hosts that are allowed unlimited connections.
// The value of RATELIMIT must be a list of the form "N X host [host ...]",
// where the first 2 numbers indicate the rate limit for unknown origins, and
// the list of hosts is the whitelist (hosts without limits).  The numbers N
// and X indicate N requests per X minutes for connections that are *not*
// coming from the listed hosts, Hosts can be regular expressions of the
// JavaScript variety, set off by forward slashes.  For example, "0 1
// /.*\.caltech\.edu/" means 0 connections per minute are allowed by default,
// except for conections from *.caltech.edu.  Note that this only affects
// actual CORS proxying; connections to the help page are *not* affected by
// this at all, and are unrestricted.  (Corrolary: don't use connections to
// the help page to judge whether this setting is working as desired.)
var rateLimit = process.env.RATELIMIT || '30 1';

// Space-separated list of header(s) that must be present in HTTP request in
// order for corsproxy to accept it.
var requiredHeader = process.env.REQUIRED_HEADER || '';

// Space-separated whitelist or blacklist for Origin.
// If the whitelist is empty or unspecified, all origins are accepted
// (except those listed explicitly in the blacklist).
var originWhitelist = process.env.ORIGIN_WHITELIST || '';
var originBlacklist = process.env.ORIGIN_BLACKLIST || '';

// File with text to override the default CORS Anywhere help text.
var helpTextFile = path.join(__dirname, 'help-text.txt');

// Main code.
// ............................................................................

// Split up space-separated fields into arrays.
function spaceSeparated(x) {
    x = x.trim();
    if (x) {
        return x.split(/\s+/g);
    } else {
        return [];
    }
}
requiredHeader = spaceSeparated(requiredHeader);
originWhitelist = spaceSeparated(originWhitelist);
originBlacklist = spaceSeparated(originBlacklist);

process.on('SIGTERM', () => {
    console.log('Received SIGTERM; exiting.');
    process.exit(0);
});

if (key_file && cert_file) {
    console.log('Using HTTPS certificate', cert_file);
    httpsOptions = {
        key: fs.readFileSync(key_file),
        cert: fs.readFileSync(cert_file)
    };
} else
    httpsOptions = {};


try {
    cors_anywhere.createServer({
        requireHeader      : requiredHeader,
        originWhitelist    : originWhitelist,
        originBlacklist    : originBlacklist,
        setHeaders         : {"X-Proxied-by": "CORS Proxy"},
        checkRateLimit     : checkRateLimit(rateLimit),
        redirectSameOrigin : true,
        httpsOptions       : httpsOptions,
        helpFile           : helpTextFile,
    }).listen(port, host, function() {
        console.log('Running CORS Proxy on ' + host + ':' + port);
        console.log('Ratelimit configuration: "' + rateLimit + '"');
        console.log('Required headers: ' + JSON.stringify(requiredHeader));
        console.log('Origin whitelist: ' + JSON.stringify(originWhitelist));
        console.log('Origin blacklist: ' + JSON.stringify(originBlacklist));
    });
} catch (err) {
    console.log(err);
    process.exit(1);
}
